File: content/06-linked-list/reverse_linked_list.js
/**
 * Problem: Reverse Linked List
 * LeetCode: #206 (Easy) | NeetCode 150 | Blind 75
 * 
 * Time Complexity: O(N)
 * Space Complexity: O(1)
 */

class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

function reverseList(head) {
    let prev = null, curr = head;
    while (curr) {
        const nxt = curr.next;
        curr.next = prev;
        prev = curr;
        curr = nxt;
    }
    return prev;
}

function toArray(head) {
    const res = [];
    while (head) {
        res.push(head.val);
        head = head.next;
    }
    return res;
}

if (require.main === module) {
    const node3 = new ListNode(3);
    const node2 = new ListNode(2, node3);
    const node1 = new ListNode(1, node2);

    const reversedHead = reverseList(node1);
    console.assert(JSON.stringify(toArray(reversedHead)) === JSON.stringify([3, 2, 1]), "Test Failed");
    console.log("All JavaScript tests passed for Reverse Linked List!");
}

File: content/06-linked-list/reverse_linked_list.py
"""
Problem: Reverse Linked List
LeetCode: #206 (Easy) | NeetCode 150 | Blind 75

Given the head of a singly linked list, reverse the list, and return the reversed list.

Time Complexity: O(N)
Space Complexity: O(1)
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverse_list(head: ListNode) -> ListNode:
    prev, curr = None, head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev

def to_list(head: ListNode) -> list:
    res = []
    while head:
        res.append(head.val)
        head = head.next
    return res

if __name__ == "__main__":
    node3 = ListNode(3)
    node2 = ListNode(2, node3)
    node1 = ListNode(1, node2)
    
    reversed_head = reverse_list(node1)
    assert to_list(reversed_head) == [3, 2, 1]
    print("All Python tests passed for Reverse Linked List!")
